File: advent/2021/23.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"os"
	"time"
)

// There are 23 valid cells, so let's get all of them indexed
var cellPosX = [23]int{0, 1, 3, 5, 7, 9, 10, 2, 2, 2, 2, 4, 4, 4, 4, 6, 6, 6, 6, 8, 8, 8, 8}
var cellPosY = [23]int{0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4}

type State struct {
	// Each element represents the state of a cell. A, B, C, D indicate the amphipods
	// while the period mark indicates empty cell.
	cells []byte
	// Total cost to reach this state from initial state
	cost int
}

// Info about one amphipod's move
type Move struct {
	// Indices of cells where amphipod stays before and after moving
	orig, dest int
	// Cost of a single move (not to be confused with total cost in State.cost)
	cost int
}

// Weighted costs by amphipod's kind
var costs = map[byte]int{
	'A': 1,
	'B': 10,
	'C': 100,
	'D': 1000,
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

// Returns information about one amphipod's move
func (s *State) getMove(index, destX, destY int) Move {
	kind := s.cells[index]
	x, y := cellPosX[index], cellPosY[index]
	distance := abs(x-destX) + abs(y-destY)
	cost := distance * costs[kind]
	destIndex := getCellIndex(destX, destY)
	return Move{index, destIndex, cost}
}

// Creates the resulting state after applying one move on original state and returns it
func (s *State) applyMove(m Move) State {
	cellsTemp := make([]byte, len(s.cells))
	copy(cellsTemp, s.cells)
	cellsTemp[m.dest] = cellsTemp[m.orig]
	cellsTemp[m.orig] = '.'
	return State{cellsTemp, s.cost + m.cost}
}

// Checks if amphipod can move along the hallway starting from column x to destX
func (s *State) isHallwayOpen(x, destX, index int) bool {
	xLeft, xRight := x, destX
	if destX < x {
		xLeft, xRight = destX, x
	}
	for i, cell := range s.cells {
		x, y := cellPosX[i], cellPosY[i]
		if i != index && y == 0 && x >= xLeft && x <= xRight && cell != '.' {
			return false
		}
	}
	return true
}

var digits = map[byte]int64{
	'.': 0,
	'A': 1,
	'B': 2,
	'C': 3,
	'D': 4,
}

// Returns the representation of cells attribute in hashable data type
// Since the cell state can only be one of 5 values (A, B, C, D, empty cell), the cells attribute
// can be represented in a base 5 system, of which the largest value (5 pow 23 - 1 ~ 1.192e16) is
// within int64 range
func (s *State) getCellsKey() int64 {
	var key int64
	for _, cell := range s.cells {
		key = 5*key + digits[cell]
	}
	return key
}

func getCellIndex(x, y int) int {
	for i := 0; i < 23; i++ {
		if cellPosX[i] == x && cellPosY[i] == y {
			return i
		}
	}
	return -1
}

// Returns a cell kind from coordinate, if a cell is invalid or out of range, `false` is returned
func (s *State) getCell(x, y int) (byte, bool) {
	index := getCellIndex(x, y)
	if index == -1 {
		return 0, false
	}
	return s.cells[index], true
}

func (s *State) amphipodCount() int {
	count := 0
	for _, cell := range s.cells {
		if cell != '.' {
			count++
		}
	}
	return count
}

var destRoom = map[byte]int{
	'A': 2,
	'B': 4,
	'C': 6,
	'D': 8,
}

func (s *State) availableMoves() []Move {
	moves := []Move{}

	maxY := 4
	if s.amphipodCount() == 8 {
		maxY = 2
	}

	for i, cell := range s.cells {
		// ignore empty cells
		if cell == '.' {
			continue
		}

		curX, curY := cellPosX[i], cellPosY[i]
		destX := destRoom[cell]

		// already in its right place
		if curX == destX && curY == maxY {
			continue
		}

		// hallway to room transition (destY is the lowest empty cell in the room)
		if curY == 0 {
			destY := -1
			for y := maxY; y >= 1; y-- {
				kind, _ := s.getCell(destX, y)
				if kind == '.' {
					destY = y
					break
				}
				if kind != cell {
					break
				}
			}
			if destY == -1 {
				continue
			}

			if s.isHallwayOpen(curX, destX, i) {
				moves = append(moves, s.getMove(i, destX, destY))
			}
			continue
		}

		// amphipod is in either of the 4 side rooms
		// check if there's any amphipod above them
		isStuck := false
		for y := 1; y < curY; y++ {
			kind, _ := s.getCell(curX, y)
			if kind != '.' {
				isStuck = true
				break
			}
		}
		if isStuck {
			continue
		}

		// suppose current amphipods is in its destination room
		// check if there's any amphipod under it that is not in its correct room
		if curX == destX {
			isSet := true
			for y := curY + 1; y <= maxY; y++ {
				kind, _ := s.getCell(curX, y)
				if kind != cell {
					isSet = false
					break
				}
			}
			if isSet {
				continue
			}
		}

		// room to hallway transition
		for _, hallX := range []int{0, 1, 3, 5, 7, 9, 10} {
			if s.isHallwayOpen(curX, hallX, i) {
				moves = append(moves, s.getMove(i, hallX, 0))
			}
		}
	}
	return moves
}

// Two states are considered equal when the cell states of both are the same
func (s *State) isEqualTo(other State) bool {
	if len(s.cells) != len(other.cells) {
		return false
	}
	for i := 0; i < len(s.cells); i++ {
		if s.cells[i] != other.cells[i] {
			return false
		}
	}
	return true
}

// Priority queue implementation
type PQueue []State

func (h PQueue) Len() int           { return len(h) }
func (h PQueue) Less(i, j int) bool { return h[i].cost < h[j].cost }
func (h PQueue) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *PQueue) Push(x any)        { *h = append(*h, x.(State)) }

func (h *PQueue) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func dijkstra(init, end []byte) (int, []State) {
	costs := make(map[int64]int)
	prev := make(map[int64]State)

	initState := State{init, 0}
	endState := State{end, 0} // a dummy one to compare with other states

	pq := PQueue{initState}
	heap.Init(&pq)

	for len(pq) > 0 {
		cur := heap.Pop(&pq).(State)
		cost := cur.cost

		if cur.isEqualTo(endState) {
			path := []State{}
			for {
				path = append(path, cur)
				if cur.isEqualTo(initState) {
					break
				}
				curKey := cur.getCellsKey()
				prev, found := prev[curKey]
				if !found {
					panic(curKey)
				}
				cur = prev
			}
			for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
				path[i], path[j] = path[j], path[i]
			}
			return cost, path
		}

		for _, move := range cur.availableMoves() {
			next := cur.applyMove(move)
			nextKey := next.getCellsKey()
			if c, ok := costs[nextKey]; !ok || next.cost < c {
				prev[nextKey] = cur
				costs[nextKey] = next.cost
				heap.Push(&pq, next)
			}
		}
	}
	return -1, nil
}

var final1 = []byte(".......AA..BB..CC..DD..")
var final2 = []byte(".......AAAABBBBCCCCDDDD")

func parseInput(scanner *bufio.Scanner) []byte {
	lines := []string{}
	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
	}
	if err := scanner.Err(); err != nil {
		panic(err)
	}

	cells := make([]byte, 23)
	for i := range cells {
		cells[i] = '.'
	}
	i := 7
	for _, col := range []int{3, 5, 7, 9} {
		cells[i] = lines[2][col]
		cells[i+1] = lines[3][col]
		i += 4
	}
	return cells
}

func getInputForPart2(cells []byte) []byte {
	cellsTemp := make([]byte, len(cells))
	copy(cellsTemp, cells)
	addition := [4][2]byte{{'D', 'D'}, {'C', 'B'}, {'B', 'A'}, {'A', 'C'}}
	for i, start := range []int{7, 11, 15, 19} {
		cellsTemp[start+3] = cellsTemp[start+1]
		cellsTemp[start+1] = addition[i][0]
		cellsTemp[start+2] = addition[i][1]
	}
	return cellsTemp
}

func main() {
	start1 := parseInput(bufio.NewScanner(os.Stdin))
	start2 := getInputForPart2(start1)

	startTime := time.Now()
	p1Cost, _ := dijkstra(start1, final1)
	p1Time := fmt.Sprint(time.Since(startTime))

	startTime = time.Now()
	p2Cost, _ := dijkstra(start2, final2)
	p2Time := fmt.Sprint(time.Since(startTime))

	fmt.Printf("Part 1: %d (%s elapsed)\n", p1Cost, p1Time)
	fmt.Printf("Part 2: %d (%s elapsed)\n", p2Cost, p2Time)
}

File: advent/2021/23.rs
// rewriting day 19's solution for the meme

use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap};
use std::io::{self, BufRead};
use std::time::Instant;

const CELLS_ROW: [u8; 23] = [
    0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4,
];
const CELLS_COL: [u8; 23] = [
    1, 2, 4, 6, 8, 10, 11, 3, 3, 3, 3, 5, 5, 5, 5, 7, 7, 7, 7, 9, 9, 9, 9,
];

fn cell_coord(idx: usize) -> (u8, u8) {
    (CELLS_ROW[idx], CELLS_COL[idx])
}

fn cell_index_at_coord(coord: (u8, u8)) -> Option<usize> {
    (0..23).find(|&i| CELLS_ROW[i] == coord.0 && CELLS_COL[i] == coord.1)
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Cell {
    E = 0,
    A = 1,
    B = 2,
    C = 3,
    D = 4,
}

impl From<char> for Cell {
    fn from(c: char) -> Self {
        match c {
            'A' => Cell::A,
            'B' => Cell::B,
            'C' => Cell::C,
            'D' => Cell::D,
            '.' => Cell::E,
            c => panic!("invalid char {:?}", c),
        }
    }
}

impl Cell {
    fn target_column(&self) -> u8 {
        const TARGET_COLS: [u8; 5] = [0, 3, 5, 7, 9];
        match self {
            Cell::E => panic!("guard this case yourself"),
            c => TARGET_COLS[*c as usize],
        }
    }

    fn weight(&self) -> u32 {
        const WEIGHTS: [u32; 5] = [0, 1, 10, 100, 1000];
        WEIGHTS[*self as usize]
    }
}

#[derive(Clone, Debug)]
struct State {
    cells: Vec<Cell>,
    cost: u32,
}

#[derive(Clone, Debug)]
struct Move {
    orig_idx: usize,
    dest_idx: usize,
    cost: u32,
}

impl State {
    // Since a cell can be one of 5 values (A, B, C, D, empty), the cells attribute
    // can be hashed in a base 5 system number
    fn cells_hash(&self) -> u64 {
        let mut res = 0;
        for &c in self.cells.iter() {
            res = 5 * res + c as u64;
        }
        res
    }

    fn amphipod_count(&self) -> usize {
        self.cells.iter().filter(|&&x| x != Cell::E).count()
    }

    fn get_move(&self, u_idx: usize, v: (u8, u8)) -> Option<Move> {
        assert!(u_idx < 23);
        let u = cell_coord(u_idx);
        let dist = (v.0 as i8 - u.0 as i8).abs() as u8 + (v.1 as i8 - u.1 as i8).abs() as u8;
        if let Some(dest_idx) = cell_index_at_coord(v) {
            return Some(Move {
                orig_idx: u_idx,
                dest_idx,
                cost: dist as u32 * self.cells[u_idx].weight(),
            });
        }
        None
    }

    fn apply_move(&self, m: &Move) -> State {
        let mut cells = self.cells.clone();
        cells[m.dest_idx] = cells[m.orig_idx];
        cells[m.orig_idx] = Cell::E;
        State {
            cells,
            cost: self.cost + m.cost,
        }
    }

    fn is_hallway_segment_empty(&self, idx: usize, col_a: u8, col_b: u8) -> bool {
        let col_range = col_a.min(col_b)..=col_a.max(col_b);
        for (i, &cell) in self.cells.iter().enumerate() {
            let (row, col) = cell_coord(i);
            if row == 0 && col_range.contains(&col) && i != idx && cell != Cell::E {
                return false;
            }
        }
        true
    }

    fn cell_at_coord(&self, coord: (u8, u8)) -> Option<Cell> {
        if let Some(idx) = cell_index_at_coord(coord) {
            return Some(self.cells[idx]);
        }
        None
    }

    fn available_moves(&self) -> Vec<Move> {
        let mut moves = vec![];
        let row_max = self.amphipod_count() as u8 / 4;

        'next_cell: for (i, &cell) in self.cells.iter().enumerate() {
            if cell == Cell::E {
                continue;
            }
            let (row, col) = cell_coord(i);

            if row == 0 {
                // Hallway to target room transition
                let dest_col = cell.target_column();
                let mut dest_row = 0;

                // Iterate over the target room from the bottom up to find the target cell.
                // Check if there are amphipods of other type inside the room
                // so they won't get stuck when the amphipod moves in target room
                for x in (1..=row_max).rev() {
                    match self.cell_at_coord((x, dest_col)) {
                        Some(Cell::E) => {
                            dest_row = x;
                            break;
                        }
                        Some(c) if c != cell => continue 'next_cell,
                        Some(_) => (),
                        None => panic!("how is {:?} not a cell", (x, dest_col)),
                    };
                }

                // Also, make sure the hallway segment between the amphipod and the
                // target room is empty before moving
                assert_ne!(dest_row, 0);
                if !self.is_hallway_segment_empty(i, col, dest_col) {
                    continue;
                }
                moves.push(self.get_move(i, (dest_row, dest_col)).unwrap());
            } else {
                // Room to hallway transition
                // Amphipod can move to the hallway when there are no amphipods overhead
                for x in 1..row {
                    match self.cell_at_coord((x, col)) {
                        Some(Cell::E) => (),
                        Some(_) => continue 'next_cell,
                        None => panic!("how is {:?} not a cell", (x, col)),
                    };
                }

                // If it's already in the target room, it only moves out when it has
                // to make way for amphipods underneath that are in the wrong room to
                // move to their target rooms
                if col == cell.target_column() {
                    let mut all_comrades = true;
                    for x in row + 1..=row_max {
                        match self.cell_at_coord((x, col)) {
                            Some(Cell::E) => panic!("they're not gonna bite you bro"),
                            Some(c) if c != cell => {
                                all_comrades = false;
                                break;
                            }
                            Some(_) => (),
                            None => panic!("how is {:?} not a cell", (x, col)),
                        }
                    }
                    if all_comrades {
                        continue 'next_cell;
                    }
                }

                const HALLWAY_COLS: [u8; 7] = [1, 2, 4, 6, 8, 10, 11];
                for &y in HALLWAY_COLS.iter() {
                    if self.is_hallway_segment_empty(i, col, y) {
                        moves.push(self.get_move(i, (0, y)).unwrap());
                    }
                }
            }
        }
        moves
    }

    fn equal_cells_with(&self, other: &Self) -> bool {
        self.cells
            .iter()
            .zip(other.cells.iter())
            .all(|(a, b)| a == b)
    }
}

impl PartialEq for State {
    fn eq(&self, other: &Self) -> bool {
        self.cost == other.cost
    }
}

impl Eq for State {}

impl Ord for State {
    fn cmp(&self, other: &Self) -> Ordering {
        other.cost.cmp(&self.cost)
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn dijkstra(init: &State, end: &State) -> Option<(Vec<Move>, u32)> {
    let mut costs: HashMap<u64, u32> = HashMap::new();
    let mut prevs: HashMap<u64, (u64, Move)> = HashMap::new();
    let mut queue = BinaryHeap::new();
    queue.push(init.clone());

    while !queue.is_empty() {
        let state = queue.pop().unwrap();
        let (cost, state_key) = (state.cost, state.cells_hash());

        // end state reached, trace move path and return total cost
        if state.equal_cells_with(end) {
            let mut path = vec![];
            let init_key = init.cells_hash();
            let mut k = state_key;
            while k != init_key {
                let (h, m) = prevs.get(&k).unwrap();
                path.push(m.clone());
                k = *h;
            }
            path.reverse();
            return Some((path, cost));
        }

        let moves = state.available_moves();
        for m in moves.iter() {
            let next = state.apply_move(m);
            let next_key = next.cells_hash();
            match costs.get(&next_key) {
                Some(&c) if next.cost >= c => continue,
                _ => {
                    costs.insert(next_key, next.cost);
                    prevs.insert(next_key, (state_key, m.clone()));
                }
            };
            queue.push(next);
        }
    }
    None
}

fn main() {
    use Cell::*;

    let mut lines = io::stdin().lock().lines().map(Result::unwrap);
    lines.next();
    let mut grid: Vec<Vec<_>> = lines.map(|l| l.chars().collect()).collect();

    let cells: Vec<_> = (0..23)
        .map(|i| {
            let (x, y) = cell_coord(i);
            if x == 1 || x == 2 {
                Cell::from(grid[x as usize][y as usize])
            } else {
                E
            }
        })
        .collect();
    let init1 = State { cells, cost: 0 };
    let end1 = State {
        cells: vec![
            E, E, E, E, E, E, E, A, A, E, E, B, B, E, E, C, C, E, E, D, D, E, E,
        ],
        cost: 0,
    };

    // add 2 more lines for part 2
    grid.insert(2, "  #D#C#B#A#".chars().collect());
    grid.insert(3, "  #D#B#A#C#".chars().collect());

    let cells: Vec<_> = (0..23)
        .map(|i| {
            let (x, y) = cell_coord(i);
            Cell::from(grid[x as usize][y as usize])
        })
        .collect();
    let init2 = State { cells, cost: 0 };
    let end2 = State {
        cells: vec![
            E, E, E, E, E, E, E, A, A, A, A, B, B, B, B, C, C, C, C, D, D, D, D,
        ],
        cost: 0,
    };

    let start = Instant::now();
    let (_p, c) = dijkstra(&init1, &end1).unwrap();
    let p1_time = start.elapsed();
    // for m in p.iter() {
    //     println!("{:?} -> {:?} ", cell_coord(m.orig_idx), cell_coord(m.dest_idx));
    // }
    println!("Part 1: {} ({}ms elapsed)", c, p1_time.as_millis());

    let start = Instant::now();
    let (_p, c) = dijkstra(&init2, &end2).unwrap();
    let p2_time = start.elapsed();
    // for m in p.iter() {
    //     println!("{:?} -> {:?} ", cell_coord(m.orig_idx), cell_coord(m.dest_idx));
    // }
    println!("Part 2: {} ({}ms elapsed)", c, p2_time.as_millis());
}
